models/seeds.py

```diff
@@ -0,0 +1,180 @@
+        'A' + rev_compl(miRNA[1:7]) # 7mer-A1 - full complementarity on positions 2-7 and A on the position 1
+    ]
+
+def seeds_6mer(miRNA):
+    return [
+        rev_compl(miRNA[1:7]), # 6mer - full complementarity on positions 2-7
+        rev_compl(miRNA[2:8]), # 6mer-m8 - full complementarity on positions 3-8
+        'A' + rev_compl(miRNA[1:6]) # 6mer-A1 - full complementarity on positions 2-6 and A on the position 1
+    ]
+
+def seeds_6mer_bulge(miRNA):
+    mers = []
+    mers.append(rev_compl(miRNA[1:7]))
+    for pos in range(1, 7):
+        for nt in ['A', 'C', 'G', 'T']:
+            mers.append(
+                rev_compl(miRNA[1:7])[:pos] + nt + rev_compl(miRNA[1:7])[pos:]
+            )
+    mers.append(rev_compl(miRNA[2:8]))
+    for pos in range(2, 8):
+        for nt in ['A', 'C', 'G', 'T']:
+            mers.append(
+                rev_compl(miRNA[2:8])[:pos] + nt + rev_compl(miRNA[2:8])[pos:]
+            )
+    mers.append('A' + rev_compl(miRNA[1:6]))
+    for pos in range(1, 6):
+        for nt in ['A', 'C', 'G', 'T']:
+            mers.append(
+                'A' + rev_compl(miRNA[1:6])[:pos] + nt + rev_compl(miRNA[1:6])[pos:]
+            )
+
+    return list(set(mers))
+
+def seeds_6mer_bulge_or_mismatch(miRNA):
+    mers = []
+    mers.append(rev_compl(miRNA[1:7]))
+    for pos in range(1, 7):
+        for nt in ['A', 'C', 'G', 'T']:
+            # bulges
+            mers.append(
+                rev_compl(miRNA[1:7])[:pos] + nt + rev_compl(miRNA[1:7])[pos:]
+            )
+            # mismatches
+            mers.append(
+                rev_compl(miRNA[1:7])[:pos] + nt + rev_compl(miRNA[1:7])[pos+1:]
+            )
+    mers.append(rev_compl(miRNA[2:8]))
+    for pos in range(2, 8):
+        for nt in ['A', 'C', 'G', 'T']:
+            mers.append(
+                rev_compl(miRNA[2:8])[:pos] + nt + rev_compl(miRNA[2:8])[pos:]
+            )
+            mers.append(
+                rev_compl(miRNA[2:8])[:pos] + nt + rev_compl(miRNA[2:8])[pos+1:]
+            )
+    mers.append('A' + rev_compl(miRNA[1:6]))
+    for pos in range(1, 6):
+        for nt in ['A', 'C', 'G', 'T']:
+            mers.append(
+                'A' + rev_compl(miRNA[1:6])[:pos] + nt + rev_compl(miRNA[1:6])[pos:]
+            )
+            mers.append(
+                'A' + rev_compl(miRNA[1:6])[:pos] + nt + rev_compl(miRNA[1:6])[pos+1:]
+            )
+
+    return list(set(mers))
+
+
+def rev_compl(st):
+    nn = {'A': 'T', 'C': 'G', 'G': 'C', 'T': 'A'}
+    return "".join(nn[n] for n in reversed(st))
+
+
+def classify_site(gene, miRNA, get_seeds, thresh):
+
+    seeds = get_seeds(miRNA)
+    seed_counts = 0
+
+    # TODO: is it possible that one position will be counted as multiple seeds?
+    for seq in seeds:
+        seed_counts += gene.count(seq)
+        if seed_counts > thresh:
+            return 1
+        
+    return 0
+
+
+def get_seed_locations(gene, miRNA, get_seeds):
+    # TODO: allow imperfect match
```

Comment:
What do you mean be perfect match?

Reply:
By imperfect I meant the string match. Allow some kind of a loose matching. However, that is not really the right place for that, since here we want to find the specific seed strings as defined. Removing the comment